### PR TITLE
Catch Redis interruption

### PR DESCRIPTION
**Pull request by jbgriesner**
RedisError was catch in the rate limiter and in the init_cache() but not in the cache.
This PR just proposes to catch it also in the cache in order to prevent any Redis interruption.

### CODE REVIEW

File: idunn/blocks/wikipedia.py [hunk 1]
```diff
@@ -180,14 +180,23 @@ def with_cache(*args, **kwargs):
...
             if cls._connection is not False:
-                value_stored = cls._connection.get(key)
-                if value_stored is not None:
-                    result = json.loads(value_stored.decode('utf-8'))
+                try:
+                    value_stored = cls._connection.get(key)
+                except RedisError:
+                    logging.error("Got a RedisError in {}".format(f.__name__), exc_info=True)
+                    return None
                 else:
```
Comment (antoine-de) [inline idunn/blocks/wikipedia.py]: this else is useless

File: idunn/blocks/wikipedia.py [hunk 2]
```diff
@@ -180,14 +180,23 @@ def with_cache(*args, **kwargs):
             otherwise we bypass it
             """
             if cls._connection is not False:
-                value_stored = cls._connection.get(key)
-                if value_stored is not None:
-                    result = json.loads(value_stored.decode('utf-8'))
+                try:
+                    value_stored = cls._connection.get(key)
```
Comment (antoine-de) [inline idunn/blocks/wikipedia.py]: i'd prefer a `get` / `set` method in the class to hide this

File: idunn/blocks/wikipedia.py [hunk 3]
```diff
@@ -180,14 +180,23 @@ def with_cache(*args, **kwargs):
...
+                        result = json.loads(value_stored.decode('utf-8'))
+                    else:
+                        result = f(*args, **kwargs)
+                        json_result = json.dumps(result)
+                        try:
+                            cls._connection.set(key, json_result, ex=cls._expire)
+                        except RedisError:
+                            logging.error("Got a RedisError in {}".format(f.__name__), exc_info=True)
+                            return None
+                    return result
```
Comment (antoine-de) [inline idunn/blocks/wikipedia.py]: in general i'd prefer early return in this function

File: idunn/blocks/wikipedia.py (hunk 1 above)
Comment (jbgriesner) [reply idunn/blocks/wikipedia.py]: done

File: idunn/blocks/wikipedia.py (hunk 2 above)
Comment (jbgriesner) [reply idunn/blocks/wikipedia.py]: done

File: idunn/blocks/wikipedia.py [hunk 4]
```diff
@@ -199,14 +198,13 @@ def with_cache(*args, **kwargs):
...
-                else:
-                    result = f(*args, **kwargs)
-                    json_result = json.dumps(result)
-                    if not cls.set_value(f, key, json_result):
-                        return None
+                    return json.loads(value_stored.decode('utf-8'))
+                result = f(*args, **kwargs)
+                json_result = json.dumps(result)
+                if not cls.set_value(key, json_result):
+                    return None
```
Comment (antoine-de) [inline idunn/blocks/wikipedia.py]: if we can't write the response in the cache, won't we cant to return this response anyway ?

Comment (jbgriesner) [reply idunn/blocks/wikipedia.py]: indeed you're right,
done

**Pull request merged**